gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/ReevaluateDagProc.java

```diff
@@ -0,0 +1,199 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.orchestration.proc;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+
+import com.codahale.metrics.Timer;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.metrics.ServiceMetricNames;
+import org.apache.gobblin.metrics.event.TimingEvent;
+import org.apache.gobblin.runtime.api.DagActionStore;
+import org.apache.gobblin.service.ExecutionStatus;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.flowgraph.DagNodeId;
+import org.apache.gobblin.service.modules.orchestration.DagManagementStateStore;
+import org.apache.gobblin.service.modules.orchestration.DagManagerUtils;
+import org.apache.gobblin.service.modules.orchestration.task.ReevaluateDagTask;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+import org.apache.gobblin.service.monitoring.FlowStatusGenerator;
+import org.apache.gobblin.service.monitoring.JobStatus;
+import org.apache.gobblin.service.monitoring.JobStatusRetriever;
+
+
+/**
+ * An implementation for {@link DagProc} that launches a new job if there exists a job whose pre-requisite jobs are
+ * completed successfully. If there are no more jobs to run and no job is running for the Dag, it cleans up the Dag.
+ */
+@Slf4j
+public class ReevaluateDagProc extends DagProc<Optional<Dag.DagNode<JobExecutionPlan>>, Void> {
+  private final JobStatusRetriever jobStatusRetriever;
+  private final Timer jobStatusPolledTimer;
+  private final DagNodeId dagNodeId;
+  private JobStatus jobStatus;
+
+  public ReevaluateDagProc(ReevaluateDagTask reEvaluateDagTask, JobStatusRetriever jobStatusRetriever) {
+    super(reEvaluateDagTask);
+    this.jobStatusRetriever = jobStatusRetriever;
+    this.jobStatusPolledTimer = metricContext.timer(ServiceMetricNames.JOB_STATUS_POLLED_TIMER);
+    this.dagNodeId = getDagNodeId();
+  }
+
+  @Override
+  protected Optional<Dag.DagNode<JobExecutionPlan>> initialize(DagManagementStateStore dagManagementStateStore)
+      throws IOException {
+    Optional<Dag.DagNode<JobExecutionPlan>> dagNode = dagManagementStateStore.getDagNode(this.dagNodeId);
+    if (!dagNode.isPresent()) {
+      log.error("DagNode not found for a Reevaluate DagAction with dag node id {}", this.dagNodeId);
```

Comment:
dag nodes should not be cleaned up without ReevaluateDagProc.act() get to run.

Reply:
if MALA malfunctions and assigned some instance to run the task even when the task was completed by other instance, it may not find the dag node (because it was cleaned up by the first lease holder)

Reply:
let's suggest this in the error message itself, and, if appropriate, include any more thorough description as a code comment, which a maintainer would find when looking to see where in the code emitted that error message